presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java

```diff
@@ -234,15 +237,49 @@ public FunctionRegistry(TypeManager typeManager, BlockEncodingSerde blockEncodin
         this.typeManager = requireNonNull(typeManager, "typeManager is null");
         this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
 
-        specializedFunctionCache = CacheBuilder.newBuilder()
+        specializedScalarCache = CacheBuilder.newBuilder()
                 .maximumSize(1000)
-                .build(new CacheLoader<SpecializedFunctionKey, FunctionInfo>()
+                .build(new CacheLoader<SpecializedFunctionKey, ScalarFunctionImplementation>()
                 {
                     @Override
-                    public FunctionInfo load(SpecializedFunctionKey key)
+                    public ScalarFunctionImplementation load(SpecializedFunctionKey key)
                             throws Exception
                     {
-                        return key.getFunction().specialize(key.getBoundTypeParameters(), key.getArity(), typeManager, FunctionRegistry.this);
+                        // TODO the function map should be updated, so that this cast can be removed
```

Comment:
Rather than have all these down casts, we should probably update the `FunctionMap` to have a `get(QualifedName, FunctionKind)` method that returns the appropriate interface